src/main/java/com/mongodb/kafka/connect/source/schema/BsonDocumentToSchema.java

```diff
@@ -111,6 +111,90 @@ private static Schema inferSchema(final String fieldPath, final BsonValue bsonVa
     }
   }
 
+  private static Schema combine(final Schema firstSchema, final Schema secondSchema) {
+    if (isSentinelValueSet(firstSchema)) {
+      return secondSchema;
+    }
+
+    if (firstSchema.equals(secondSchema)) {
+      return firstSchema;
+    }
+
+    if (firstSchema.type() != secondSchema.type()) {
+      LOGGER.debug(
+          "Can't combine non-matching schema types: {} and {}",
+          firstSchema.type(),
+          secondSchema.type());
+      return DEFAULT_INFER_SCHEMA_TYPE;
+    }
+
+    if (firstSchema.type() != Schema.Type.STRUCT || secondSchema.type() != Schema.Type.STRUCT) {
+      LOGGER.debug("Can't combine non-equal schema that are not both structs");
+      return DEFAULT_INFER_SCHEMA_TYPE;
+    }
+
+    SchemaBuilder builder = SchemaBuilder.struct().name(firstSchema.name()).optional();
+
+    // _id field first
+    Field _id1 = firstSchema.field(ID_FIELD);
+    Field _id2 = secondSchema.field(ID_FIELD);
+    if (_id1 != null || _id2 != null) {
+      builder.field(ID_FIELD, combineFieldSchema(_id1, _id2));
+    }
+    // Combine other fields in name order
+    Stream.concat(
+            firstSchema.fields().stream().map(Field::name),
+            secondSchema.fields().stream().map(Field::name))
+        .filter(name -> !name.equals(ID_FIELD))
+        .sorted()
```

Comment:
We can squeeze a bit more efficiency out of this by moving the call to `sorted` after the call to `distinct`, as it will halve (at best) the number of elements needed to be sorted

Reply:
Good catch - done :)